source/common/http/http1/conn_pool.cc

```diff
@@ -273,11 +275,15 @@ ConnPoolImpl::StreamWrapper::~StreamWrapper() {
 void ConnPoolImpl::StreamWrapper::onEncodeComplete() { encode_complete_ = true; }
 
 void ConnPoolImpl::StreamWrapper::decodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
-  if (headers->Connection() &&
-      absl::EqualsIgnoreCase(headers->Connection()->value().getStringView(),
-                             Headers::get().ConnectionValues.Close)) {
-    saw_close_header_ = true;
-    parent_.parent_.host_->cluster().stats().upstream_cx_close_notify_.inc();
+  if (headers->Connection()) {
+    if (absl::EqualsIgnoreCase(headers->Connection()->value().getStringView(),
+                               Headers::get().ConnectionValues.Close)) {
+      saw_close_header_ = true;
+      parent_.parent_.host_->cluster().stats().upstream_cx_close_notify_.inc();
+    } else if (absl::EqualsIgnoreCase(headers->Connection()->value().getStringView(),
+                                      Headers::get().ConnectionValues.KeepAlive)) {
+      saw_keep_alive_header_ = true;
```

Comment:
Optional but in the http connection manager we treat lack of keepalive for HTTP/1.0 as equivalent to explicit connection: close for HTTP/1.1 (i.e. use the same boolean) to make sure anywhere we special case behavior for one we special case it for the other.

I'm fine with you using separate booleans but I do think you should increment the upstream_cx_close_notify_ stat for this case (as it already is reused for GOAWAY) at which point you have to do protocol checks in this function at which point it might be nice avoiding the complexity in onResponseComplete

Reply:
Concretely are you suggesting combining the booleans into saw_upstream_close_ or some similar which is either Connection: close or for HTTP/1.0 the lack of the keep-alive header ?

Conversely I could move the upstream_cx_close_notify_ into onResponseComplete.

Reply:
The protocol isn't available in decodeHeaders() so I moved the stat increment into onResponseComplete which is consistent with the log message and the increment of upstream_cx_max_requests_.   PTAL

Reply:
+1 to Alyssa's suggestion of having a single boolean track whether we want to close the connection after processing headers.

Reply:
The protocol isn't available in decodeHeaders().  Without it I can't have a single variable.  Suggestions?

Reply:
You can get the `protocol()` from the codec, which is updated prior to header callbacks.

/wait

Reply:
Done.